### PR TITLE
Port Versioning Functionality

### PR DESCRIPTION
**Pull request by h0tl33t**
This PR ports versioning functionality from our existing CMSs, makes it app agnostic, and adds 100% test coverage.

### CODE REVIEW

File: lib/active_versioning/model.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+          versionable.to_s,
+          committed_at.try(:utc).try(:strftime, '#%Y%m%d%H%M%S')
+        ].compact.join(' ')
+      end
+
+      def reify
+        resource = versionable(true)
+
+        # Reload the versionable so we get a new one
+        versionable(true)
```
Comment (efatsi) [inline lib/active_versioning/model.rb]: `versionable(true)` twice?

File: lib/active_versioning/model/version_proxy.rb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        @record
+      end
+
+      def __setobj__(record)
+        @record = record
+      end
+
+      # `class` should delegate to versioned resource, not return
+      # ActiveVersioning::Model::VersionProxy
+      undef class
```
Comment (efatsi) [inline lib/active_versioning/model/version_proxy.rb]: :cool:

File: lib/active_versioning/model.rb (hunk 1 above)
Comment (h0tl33t) [inline lib/active_versioning/model.rb]: This ensures that the `resource` in this case is a different Ruby object than the version's `versionable`.  The result of this method is the object that's modified when you're working with a record's `current_draft`.  Might encounter bugginess if a record and its `current_draft` (just a delegator) point to the same record in memory?  @yaychris can probably explain/confirm as to why this is necessary better than I.

File: spec/lib/active_versioning/model/versioned_spec.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    end
+  end
+
+  describe "#version_author" do
+    it "should be able to set and view the author" do
+      subject.version_author = user
+
+      expect(subject.version_author).to eq user
+    end
+  end
```
Comment (efatsi) [inline spec/lib/active_versioning/model/versioned_spec.rb]: this confuses me a bit, it's just an `attr_accessor`? what good does it do?

File: spec/lib/active_versioning/model/versioned_spec.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  let!(:user) { ActiveVersioning::Test::User.create(name: 'Bo Janglez', email: 'bo.janglez@gmail.com') }
+
+  subject { ActiveVersioning::Test::Post.create(title: 'So Post', body: 'Such interesting.  Very wow.', user: user) }
+
+  it { should have_many(:versions) }
+
+  describe "after create" do
+    it "creates a new version" do
+      expect {
+        ActiveVersioning::Test::Post.create(title: 'So Post', body: 'Such interesting.  Very wow.', user: user, version_author: user)
```
Comment (efatsi) [inline spec/lib/active_versioning/model/versioned_spec.rb]: you could use `build` to instantiate the subject and just call `subject.save` here if you want to clean things up

File: spec/lib/active_versioning/model/versioned_spec.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    end
+
+    context "when there are no draft versions" do
+      it "creates a new draft version wrapped in a VersionProxy" do
+        expect {
+          subject.current_draft
+
+          # Verify we have a VersionProxy
+          expect(subject.current_draft.__getobj__).to eq subject
+        }.to change { subject.versions.count }.by 1
```
Comment (efatsi) [inline spec/lib/active_versioning/model/versioned_spec.rb]: would like to see this split into two tests, one for the new draft version, one testing the draft's `__getobj__` is the proper object

File: spec/lib/active_versioning/model/versioned_spec.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  describe "#create_draft_from_version" do
+    let!(:version) { subject.current_draft }
+
+    before { version.commit }
+
+    it "creates a new draft from the given version" do
+      expect {
+        subject.create_draft_from_version(version.id)
+      }.to change { subject.versions.count }.by 1
```
Comment (efatsi) [inline spec/lib/active_versioning/model/versioned_spec.rb]: this test is really just checking that a new version is created, is there anything else worth checking to ensure that the draft was created from the given version?

File: spec/lib/active_versioning/model/versioned_spec.rb (hunk 3 above)
Comment (h0tl33t) [inline spec/lib/active_versioning/model/versioned_spec.rb]: When a versioned object is saved, the `version_author` is pulled off the record and stored with the version as the `committer`.  Yeah, this test is super simple and just makes sure the necessary methods are still there.

File: spec/lib/active_versioning/model/versioned_spec.rb (hunk 6 above)
Comment (h0tl33t) [inline spec/lib/active_versioning/model/versioned_spec.rb]: I'll add a test that verifies the new version's object matches that of the given version.

File: spec/lib/active_versioning/model_spec.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    it { expect(subject.to_s).to match(/20141126124235$/) }
+  end
+
+  describe "#reify" do
+    context "with a valid object" do
+      let!(:version) { subject.reify }
+
+      it "returns a different Ruby object than versionable" do
+        expect(version.object_id).to_not eq(subject.versionable.object_id)
+      end
```
Comment (efatsi) [inline spec/lib/active_versioning/model_spec.rb]: this what you get from that double `(true)` stuff?

Comment (h0tl33t) [inline spec/lib/active_versioning/model_spec.rb]: Yup.

File: spec/lib/active_versioning/model_spec.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    context "with an object containing unknown attributes" do
+      let!(:version) do
+        ActiveVersioning::Test::Version.create(
+          versionable: post,
+          object:      { 'title' => 'Ruh Roh', 'rating' => '4 Stars' },
+        )
+      end
+
+      it "doesn't raise an unknown attribute error" do
```
Comment (efatsi) [inline spec/lib/active_versioning/model_spec.rb]: thought here being you might have attributes you don't want to version?

File: lib/active_versioning/model.rb (hunk 1 above)
Comment (yaychris) [inline lib/active_versioning/model.rb]: Your explanation is correct. If you don't do this, `resource` and `versionable` are the same object, which causes weird problems.

File: spec/lib/active_versioning/model_spec.rb (hunk 8 above)
Comment (h0tl33t) [inline spec/lib/active_versioning/model_spec.rb]: This is a stopgap so we don't get application errors here.  I'm working on a better overall solution for handling outdated versions that'll come in a later PR.

**Comment by efatsi** [comment]
:+1:

File: README.md [hunk 9]
```diff
@@ -1,8 +1,6 @@
 # ActiveVersioning
 
-Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/active_versioning`. To experiment with that code, run `bin/console` for an interactive prompt.
-
-TODO: Delete this and the text above, and describe your gem
+ActiveVersioning provides out-of-the-box versioning functionality in Rails.  ActiveVersioning serializes attributes when records are saved and allows for version and draft management.
```
Comment (mackermedia) [inline README.md]: Looks like some of the code examples in the README show using `current_admin`. Is this thing expected to be an enhancement only when using Active Admin? If so, might be worth noting in the intro.

File: README.md [hunk 10]
```diff
@@ -20,17 +18,95 @@ Or install it yourself as:
...
+If you require additional versioned attributes, overwrite the `versioned_attribute_names` method in your model:
+```ruby
+class Post < ActiveRecord::Base
+  private
+
+  def versioned_attribute_names
+   super + %w(photo_id)
+  end
+end
+```
```
Comment (mackermedia) [inline README.md]: Excellent README yo

File: README.md (hunk 9 above)
Comment (h0tl33t) [inline README.md]: I think my plan is to have all ActiveAdmin-specific functionality available in a separate gem where `ActiveVersioning` is a dependency.

I'll take out references to `current_admin`.

File: README.md (hunk 10 above)
Comment (Fosome) [inline README.md]: :+1: :README:

File: lib/active_versioning/model/versioned.rb [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+        @current_draft ||= begin
+          # FIXME: Gracefully handle NoMethodError and UnknownAttributeError
+          # If past versions contain renamed/different attributes, this causes problems
+          # Detect such attributes and warn appropriately.
+          # Create a nice errors object that encapsulates this information that can be used to diff.
+
+          VersionProxy.new(versions.draft.first_or_create(
+            object: versioned_attributes,
+            event:  'draft'
```
Comment (Fosome) [inline lib/active_versioning/model/versioned.rb]: use constant?

Comment (h0tl33t) [inline lib/active_versioning/model/versioned.rb]: YES.  I'm definitely going to do this.  Probably need to add some validations around the acceptable events as well.  :+1:

File: spec/active_versioning_spec.rb [hunk 12]
```diff
@@ -1,11 +1,7 @@
...
 
 describe ActiveVersioning do
-  it 'has a version number' do
-    expect(ActiveVersioning::VERSION).not_to be nil
-  end
-
-  it 'does something useful' do
-    expect(false).to eq(true)
+  it "has the correct version number" do
+    expect(ActiveVersioning::VERSION).to eq '0.1.0'
```
Comment (Fosome) [inline spec/active_versioning_spec.rb]: Necessary test?

**Comment by Fosome** [comment]
:+1:

**Comment by h0tl33t** [comment]
Much thanks, guys!

**Pull request merged**